### ISSUE TITLE
Chunk Sort Order Lost on Restart

### ISSUE THREAD

**Issue by tustvold** [influxdata/influxdb_iox#3845]
**Problem**

Create a database and persist some data

```
$ cargo run -- database create org_bucket
$ cargo run -- database write org_bucket ~/Downloads/data.lp
$ cargo run -- database partition persist org_bucket "2021-04-26 13:00:00" --all-tables --force
```

Data has a sort ordinal

```
org_bucket> select * from system.chunk_columns;
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| partition_key       | chunk_id                             | table_name | column_name | storage         | row_count | null_count | min_value           | max_value           | memory_bytes | chunk_order | sort_ordinal |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag0        | ObjectStoreOnly | 10000     | 0          | value0              | value1              |              | 1           | 1            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag1        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 2            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag2        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 3            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag3        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 4            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag4        | ObjectStoreOnly | 10000     | 0          | value0              | value0              |              | 1           | 0            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | time        | ObjectStoreOnly | 10000     | 0          | 1619444823731174000 | 1619444859727574000 |              | 1           | 5            |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
```

Unload chunk

```
$ cargo run -- database partition unload-chunk org_bucket "2021-04-26 13:00:00" "m0" "b257e6ec-dfa3-4690-a243-4cf0b852a41a"
```

Chunk still has sort ordinal

```
org_bucket> select * from system.chunk_columns;
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| partition_key       | chunk_id                             | table_name | column_name | storage         | row_count | null_count | min_value           | max_value           | memory_bytes | chunk_order | sort_ordinal |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag0        | ObjectStoreOnly | 10000     | 0          | value0              | value1              |              | 1           | 1            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag1        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 2            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag2        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 3            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag3        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           | 4            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag4        | ObjectStoreOnly | 10000     | 0          | value0              | value0              |              | 1           | 0            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | time        | ObjectStoreOnly | 10000     | 0          | 1619444823731174000 | 1619444859727574000 |              | 1           | 5            |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | v0          | ObjectStoreOnly | 10000     | 0          | 1                   | 1                   |              | 1           |              |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
```

Restart database

```
$ cargo run -- database restart org_bucket
```

Data no longer has sort ordinal

```
org_bucket> select * from system.chunk_columns;
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| partition_key       | chunk_id                             | table_name | column_name | storage         | row_count | null_count | min_value           | max_value           | memory_bytes | chunk_order | sort_ordinal |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag0        | ObjectStoreOnly | 10000     | 0          | value0              | value1              |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag1        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag2        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag3        | ObjectStoreOnly | 10000     | 0          | value0              | value9              |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | tag4        | ObjectStoreOnly | 10000     | 0          | value0              | value0              |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | time        | ObjectStoreOnly | 10000     | 0          | 1619444823731174000 | 1619444859727574000 |              | 1           |              |
| 2021-04-26 13:00:00 | b257e6ec-dfa3-4690-a243-4cf0b852a41a | m0         | v0          | ObjectStoreOnly | 10000     | 0          | 1                   | 1                   |              | 1           |              |
+---------------------+--------------------------------------+------------+-------------+-----------------+-----------+------------+---------------------+---------------------+--------------+-------------+--------------+
```

Possibly relates to #3214

**Comment by e-dard** [comment]
As if by magic! :-)

**Comment by tustvold** [comment]
**Problem**

IOx stores the sort key as schema metadata associated with the `CatalogChunk`.

`ReorgPlanner::compact_plan` and `ReorgPlanner::split_plan` return a `Schema` along with a `LogicalPlan`. This schema has a sort key set, however, the `LogicalPlan` and by extension the produced `RecordBatch` do not.

This schema is upserted into the in-memory catalog, and stored in the `ChunkMetadata`.

The `Deduplicator` inspects this schema and its corresponding sort key when generating scan plans.

The `system.column_chunks` system table also uses the data from the schema stored in the in-memory catalog.

However, the underlying `RBChunk` is not aware of this schema and its sort key.

The write lifecycle action calls `DbChunk::read_filter` on the provided chunk, which returns a stream of `RecordBatch` that are then written to a parquet file.

This data is streamed directly from the `RBChunk`, which is not sort aware, and therefore the produced `ParquetChunk` is similarly not sort aware.

On restart, the `CatalogChunk` are recreated from the metadata of the persisted parquet files, which is stored in the `PreservedCatalog`. 

As already established, the `ParquetChunk` are not sort aware, and so the recreated `CatalogChunk` also lacks sort information

**Proposal**

In hindsight storing the sort key as part of the schema was probably a mistake:

* It can lead to subtle bugs where something unaware of the SortKey reorders the records within a batch, but blindly copies across the schema metadata
* Codepaths can easily accidentally omit encoding the sort key, as it is an implicit part of the schema, as opposed to an explicit field
* https://github.com/apache/arrow-datafusion/pull/1776 added a different way to express sorted-ness to DataFusion
* It confuses the notion of schema-compatibility, are two schemas compatible if they have different metadata? What if the difference is sort order?
* There have been many upstream bugs due to operators incorrectly discarding field metadata - https://github.com/apache/arrow-datafusion/issues/1361 https://github.com/apache/arrow-datafusion/issues/1476

I would therefore like to propose making the sort order an explicit property of a chunk, and no longer serialize it as arrow metadata. Specifically it would be stored in `catalog::ChunkMetadata` and serialized to the `IOxMetadata` payload written to parquet.

Sort order would still be expressed using the existing `SortKey` and `ColumnSort` structures, but instead of being stored on a chunk's schema, it would be a property of the chunk itself.

**Comment by e-dard** [comment]
This makes sense to me.

**Comment by e-dard** [comment]
I guess one question is how we/whether we modify the sort key based on projections or deletes. For example if we delete all data in a column for a chunk whether we include that column in the sort key when we compact it etc.

**Comment by crepererum** [comment]
Sounds good to me.

I just want to point out another potential fix for this issue: Either fixing the fact that the RB isn't aware of the sort order or side-injecting the schema back at the right point since `RecordBatchStream` provides us a way to specify a schema that is different (e.g. has more information) than the schema of the actual batches.

**Comment by tustvold** [comment]
> For example if we delete all data in a column for a chunk whether we include that column in the sort key when we compact it etc.

If the tag column exists, it will need to continue to be included in the sort key even if it is still nulls. If the column was removed, which I don't think can currently happen, then it would need to also be removed from the sort key

> side-injecting the schema back at the right point since RecordBatchStream provides us a way to specify a schema that is different (e.g. has more information) than the schema of the actual batches.

I'm not sure this will work, and if it does it would more work by accident. DataFusion makes pretty strong assumptions that the schema returned by the operator, and by extension the stream, matches that of the batches. You could end up with some pretty wild bugs if you violated that...

> fixing the fact that the RB isn't aware of the sort order

I hope to do this in future, but I don't want this to uncontrollably leak out into Datafusion or similar where you might end up with a RecordBatch claiming to be sorted, that isn't

**Comment by tustvold** [comment]
Closed by #3864
